m/dram.1.in
.\" vim: tw=72 fdm=marker cms=.\\"\ %s
.
.\" FRONT MATTER {{{
.Dd December 13, 2022
.Dt DRAM 1
.Os
.
.Sh NAME
.Nm dram
.Nd literate functional tests for the CLI
.\" FRONT MATTER }}}
.
.\" SYNOPSIS {{{
.Sh SYNOPSIS
.Nm
. Fl h | V
.Nm
. Op Fl DETUfiuv
. Op Fl I Ar INDENT
. Oo Fl e Ar VAR Ns Oo Li = Ns Ar VAL Oc Oc Ns ...
. Op Fl s Ar SHELL
. Op Fl t Ar SUFFIX
. Ar TEST...
.\" SYNOPSIS }}}
.
.\" DESCRIPTION {{{
.Sh DESCRIPTION
.Nm
is a tool for functional tests of CLI applications.
.Pp
.Nm
extracts all commands from a testfile into a script,
runs the script with a modified environment, in an empty
directory, with standard input redirected from
.Pa /dev/null ,
and standard output plus standard error redirected into
a file.
Individual commands are fenced from each other
so their outputs are unambiguously attributable.
.Pp
.Nm
then compares the actual outputs with expected ones,
and offers their differences (as unified diffs) for merging
into the original testfiles.
.\" }}}
.
.\" OPTIONS {{{
.Sh OPTIONS
.Bl -tag -width xx
. It Fl h
Display usage information.
. It Fl V
Display version information.
. It Fl D
Do not print diffs.
. It Fl E
Preserve existing environment variables.
. It Fl I Ar INDENT
Number of spaces used for test code indentation
. Pq default: Li 2 .
. It Fl T
Leave behind working files.
. It Fl U
Do not merge actual output into tests.
Overrides preceding
. Fl i
and
. Fl u .
. It Fl e Ar VAR Ns Op Li = Ns Ar VAL
Run tests with
. Ev VAR
set to
. Ar VAL
in environment.
In the
. Fl e Ar VAR
form,
. Ev VAR
is copied from
. Nm Ns 's
environment.
. It Fl f
Abort after first failed test.
Skip any remaining tests.
. It Fl i
Interactively merge actual output into tests.
Overrides preceding
. Fl U
and
. Fl u .
. It Fl s Ar SHELL
Shell to use for running tests
. Pq default: Li sh .
. It Fl t Ar SUFFIX
Extension of testfiles
. Pq default: Li .t .
. It Fl u
Merge actual output into tests.
Overrides preceding
. Fl U
and
. Fl i .
. It Fl v
Show filenames with test statuses.
.El
.\" }}}
.
.\" ENVIRONMENT {{{
.Sh ENVIRONMENT
.Ss Variables used by Nm
.
Options take precedence over corresponding environment variables.
.
.Bl -tag -width xx
. It Ev DRAM_DIFF
Program used to generate diffs.
Defaults to
. Pa diff .
. It Ev DRAM_ENV
Whitespace-separated list of environment variables to copy from
.Nm Ns 's
environment into the environment of each test.
. It Ev DRAM_FAIL_FAST
If set to any of
. Qq Li 1Yy ,
enables
. Fl f .
. It Ev DRAM_INDENT
String used to indent command, continuation, output, and
exit code lines.
. It Ev DRAM_KEEP_ENVIRON
If set to any of
. Qo Li 1Yy Qc , enables Fl E .
. It Ev DRAM_KEEP_TMPDIR
If set to any of
. Qo Li 1Yy Qc , enables Fl T .
. It Ev DRAM_NODIFFS
If set to any of
. Qo Li 1Yy Qc , enables Fl D .
. It Ev DRAM_PATCH
Program used to apply diffs.
Defaults to
. Pa patch .
. It Ev DRAM_SHELL
See
. Fl s .
. It Ev DRAM_UPDATE
If set to any of
. Qo Li Aa Qc , enables Fl i .
If set to any of
. Qo Li 1Yy Qc , enables Fl u .
If unset or set to any other value, enables
. Fl U .
. It Ev DRAM_TEST_SUFFIX
See
. Fl t .
. It Ev DRAM_TMPDIR
If set,
. Nm
will create working files in this directory
instead of creating one under
. Ev TMPDIR .
. It Ev DRAM_VERBOSE
If set to any of
. Qo Li 1Yy Qc , enables Fl v .
.El
.Pp
Beside the above,
.Nm
may be influenced by environment variables used by
.Xr diff 1 ,
.Xr patch 1 ,
.Xr sh 1 ,
possibly others.
.
.Ss Environment of test scripts
.
.Nm
runs tests with environment limited to the following
variables and values by default.
This behavior can be inhibited by
.Fl E
or
.Ev DRAM_KEEP_ENVIRON , and augmented using
.Fl e Ar VAR Ns Op Li = Ns Ar VAL
and/or
.Ev DRAM_ENV .
Environment variables with names starting with
.Qq Li DRAM_
are
.Em always
removed from the environment of tests.
.
.Bl -tag -width xx
. It Ev COLUMNS Ns Li = Ns Qq Li 80
. It Ev LANG Ns Li = Ns Qq Li C
. It Ev LC_ALL Ns Li = Ns Qq Li C
. It Ev LINES Ns Li = Ns Qq Li 20
. It Ev PATH Ns Li = Ns Qq Li /usr/local/bin:/usr/local/sbin:/bin:/sbin:/usr/bin:/usr/sbin .
Inherited from
. Nm Ns No 's environment if present .
. It Ev TESTDIR
Absolute path of the directory containing the testfile.
. It Ev TESTFILE
Basename of the testfile.
. It Ev TMPDIR
Pathname of an existing, writable, empty directory.
. It Ev TZ Li UTC
.El
.\" }}}
.
.\" EXIT STATUS {{{
.Sh EXIT STATUS
.Bl -tag -width xx
.It 0
Success.
.It 1
Usage error.
.It 2
A test failed.
.It 4
A test was broken.
.It 8
External utility error.
.El
.\" }}}
.
.\" EXAMPLES {{{
.Sh EXAMPLES
.Bd -literal
$ dram t
s.!X
--- 3.t
+++ 3.t
@@ -1,2 +1,2 @@
   $ echo hello
-  goodbye
+  hello

# 3 tests, 1 skipped, 1 failed, 1 broken
$ echo $?
6
.Ed
.\" }}}
.
.\" SEE ALSO {{{
.Sh SEE ALSO
.Xr diff 1 ,
.Xr patch 1 ,
.Xr sh 1 ,
.Xr dram 5 ,
.Lk https://bitheap.org/cram .
.\" }}}
.
.\" AUTHORS {{{
.Sh AUTHORS
.An Roman Neuhauser Aq Mt rn+dram@sigpipe.cz
.\" AUTHORS }}}
.
.\" BUGS {{{
.Sh BUGS
No doubt plentiful.
Please report them at
.Lk https://todo.sr.ht/~rne/dram .
.\" BUGS }}}


debian/changelog
dram (0.10-1) stable; urgency=medium

  [ Roman Neuhauser ]
  * GNUmakefile: install man page
  * m/dram.1.in: syntax fixes and changes
  * s/dram.d: correct a comment
  * m/dram.5.in
  * flush stdout after each test status in default mode
  * manpage fixes
  * dram.d: replace escapeShellCommand with custom code

  [ Jan Matějka ]
  * debian: build for trixie
  * debian/patches: gbp pq export

 -- Jan Matějka <jan@matejka.ninja>  Thu, 04 Sep 2025 11:17:15 +0200

dram (0.8-1) unstable; urgency=medium

  [ Jan Matějka ]
  * Initial release.
  * debian: gbp pq export
  * debian: add gbp.conf

 -- Jan Matějka <jan@matejka.ninja>  Tue, 21 May 2024 04:26:12 +0200


s/dram.d
// vim: et sts=2 fdm=marker cms=\ //\ %s
//
// Copyright (C) 2020  Roman Neuhauser
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

module dram;

import core.time;
import std.algorithm;
import std.conv;
import std.container.array;
import std.datetime.stopwatch: AutoStart, StopWatch;
import std.datetime.systime;
import std.file;
import std.path;
import std.process;
import std.range;
import std.regex;
import std.stdio;
import std.string;
import std.typecons;
import dxml.util;
import dxml.writer;

import mdiff;

int main(string[] argv) // {{{
{
  // stdout has a huge buffer by default, this causes
  // output on stderr to appear in the terminal way out
  // of order.  setting both streams to line-buffered
  // fixes this.
  [stdout, stderr].each!(f => f.setvbuf(1024, _IOLBF));

  // std.getopt sucks.  -h/--help handling is hardcoded,
  // defaultGetoptPrinter prints long options right-justified
  // which looks horribly messy, and an unknown option
  // causes an exception.  it's fucked up.
  //
  // until and unless i can fix this, dram is configured
  // using environment variables only.
  auto cfg = DramConfig(environment.toAA);

  try {
    // each operand is either a testfile or a directory
    // (supposedly) containing testfiles (recursively)
    auto tests = argv[1..$]
      .map!(arg => cfg.findTestFiles(arg))
      .joiner
      .map!buildNormalizedPath
      .array
    ;

    cfg.setUp;

    auto results = tests
      .map!(test => runTest(cfg, test))
      .cache
      // print dots, exclamation marks, etc.
      .tee!(r => cfg.report(r))
      // abort on first fail if given -f
      .until!(r => cfg.failFast(r))(No.openRight)
      // force evaluation, need reported results
      // before printing the newline below.
      .array
    ;

    auto total = tests.length;

    if (results.length < total) {
      auto done = results.map!(tr => tr.testFile);
      results ~= tests
        .filter!(x => !done.canFind(x))
        .map!(tf => TestResult(TestStatus.skip, tf))
        .cache
        .tee!(r => cfg.report(r))
        .array
      ;
    }

    // tally results
    auto broken = 0;
    auto fails  = 0;
    auto skips  = 0;
    results.each!((r) {
      broken += r.error;
      fails  += r.fail;
      skips  += r.skip;
    });

    if (!cfg.verbose && !results.empty)
      writeln;

    // if any test failed, exit 1
    results.each!(r => cfg.diffAndPatch(r));

    auto summary = "\n# Ran %1$d %2$s";
    if (skips)  summary ~= ", %3$d skipped";
    if (fails)  summary ~= ", %4$d failed";
    if (broken) summary ~= ", %5$d broke";
    summary ~= ".";
    summary.writefln(
      total
    , total == 1 ? "test" : "tests"
    , skips
    , fails
    , broken
    );

    cfg.cleanUp(results, skips, fails);

    auto exit = 0;
    if (broken) exit |= 0b100;
    if (fails)  exit |= 0b010;
    return exit;
  } catch (ToolFailure e) {
    stderr.writeln("Failed command: " ~ e.msg);
    return 0b1000;
  } catch (ProcessException e) {
    stderr.writeln(e.msg);
    return 0b1000;
  } catch (FileException e) {
    stderr.writeln(e.msg);
    return 0b1000;
  }
} // }}}

struct DramConfig // {{{
{
  string[] envvars;
  bool   failfast;
  bool   keepenv;
  bool   keeptmp;
  bool   nodiffs;
  bool   verbose;
  bool   update;
  bool   ask;
  string extension = ".t";
  string junitFile;

  StopWatch stopWatch;

  string indent;
  Tuple!(
    string, "ps1",
    string, "ps2",
  ) prompts;

  Tuple!(
    string, "glob",
    string, "regex",
    string, "noeol",
  ) suffixes;

  string cookie;

  string diffcmd;
  string patchcmd;
  string shell;
  string tmpdir;

  this(string[string] env) // {{{
  {
    envvars = env.get("DRAM_ENV", "").split;
    failfast = isAnyOf("1Yy", env.get("DRAM_FAIL_FAST", "0"));
    keepenv = isAnyOf("1Yy", env.get("DRAM_KEEP_ENVIRON", "0"));
    keeptmp = isAnyOf("1Yy", env.get("DRAM_KEEP_TMPDIR", "0"));
    nodiffs = isAnyOf("1Yy", env.get("DRAM_NODIFFS", "0"));
    verbose = isAnyOf("1Yy", env.get("DRAM_VERBOSE", "0"));
    update = isAnyOf("1YyAa", env.get("DRAM_UPDATE", "0"));
    ask = isAnyOf("Aa", env.get("DRAM_UPDATE", "0"));
    extension = env.get("DRAM_TEST_SUFFIX", ".t");
    junitFile = env.get("DRAM_JUNIT_FILE", "");

    stopWatch = StopWatch(AutoStart.no);

    indent = env.get("DRAM_INDENT", "  ");
    prompts = tuple(
      indent ~ "$ ",
      indent ~ "> ",
    );
    suffixes = tuple(
      " (glob)",
      " (re)",
      " (no-eol)",
    );

    cookie = "DRAM" ~ thisProcessID.to!string;

    diffcmd = env.get("DRAM_DIFF", "diff");
    patchcmd = env.get("DRAM_PATCH", "patch");
    shell = env.get("DRAM_SHELL", "sh");
    tmpdir = env.get("DRAM_TMPDIR", tempDir.buildPath("dramtests-" ~ thisProcessID.to!string));
  } // }}}
  string default_(string def, string val) // {{{
  {
    return val.empty ? def : val;
  } // }}}
  bool isAnyOf(string chars, string val) // {{{
  {
    return val.empty ? false : chars.any!(x => x == val[0]);
  } // }}}

  string[] findTestFiles(string under) // {{{
  {
    if (under.isDir)
      return dirEntries(under, SpanMode.breadth)
        .map!(to!string)
        .filter!(fn => fn.isFile && fn.extension == extension)
        .array // `sort` needs a random-access container
        .sort
        .array
      ;
    else
      return [under];
  } // }}}

  void setUp() // {{{
  {
    stopWatch.start;
    tmpdir.mkdirRecurse;
  } // }}}
  void cleanUp(TestResult[] results, int skipped, int failed) // {{{
  {
    if (!junitFile.empty)
      writeJUnitFile(results, skipped, failed);
    if (!keeptmp)
      tmpdir.rmdirRecurse;
    else
      "preserved %s".writefln(tmpdir);
  } // }}}
  void writeJUnitFile(TestResult[] results, int skipped, int failed) // {{{
  // based on https://llg.cubic.org/docs/junit/
  // deviation from Cram: omits <testsuite>'s hostname attribute
  {
    auto now = Clock.currTime!(ClockType.second);
    auto suiteTime = stopWatch.peek.total!"msecs" / 1000.0;
    auto junit = File(junitFile, "w");
    auto orange = junit.lockingTextWriter;
    orange.writeXMLDecl!string;
    auto xml = xmlWriter(orange, "  ");
    xml.openStartTag("testsuite", Newline.yes);
    xml.writeAttr("name", "dram", Newline.yes);
    xml.writeAttr("tests", results.length.to!string, Newline.yes);
    xml.writeAttr("failures", failed.to!string, Newline.yes);
    xml.writeAttr("skipped", skipped.to!string, Newline.yes);
    xml.writeAttr("timestamp", now.toISOExtString, Newline.yes);
    xml.writeAttr("time", suiteTime.to!string, Newline.yes);
    xml.closeStartTag;
    results.each!((tr) {
      auto ttime = tr.time.total!"msecs" / 1000.0;
      xml.openStartTag("testcase", Newline.yes);
      xml.writeAttr("classname", encodeAttr(tr.testFile), Newline.yes);
      xml.writeAttr("name", encodeAttr(tr.testFile.baseName), Newline.yes);
      xml.writeAttr("time", ttime.to!string, Newline.yes);
      if (tr.skip) {
        xml.closeStartTag;
        xml.openStartTag("skipped");
        xml.closeStartTag(EmptyTag.yes);
        xml.writeEndTag("testcase");
      }
      else if (tr.fail) {
        xml.closeStartTag;
        tr.diff.rewind;
        auto difftext = appender!string;
        //difftext.reserve(diff.size);
        tr.diff
          .byLine(Yes.keepTerminator)
          .each!(l => difftext ~= encodeText(l))
        ;
        xml.openStartTag("failure", Newline.yes);
        xml.closeStartTag;
        xml.writeText(difftext[], Newline.yes, InsertIndent.no);
        xml.writeEndTag("failure", Newline.yes);
        xml.writeEndTag("testcase");
      } else {
        xml.closeStartTag(EmptyTag.yes);
      }
    });
    xml.writeEndTag("testsuite");
    junit.writeln;
    junit.close;
  } // }}}
  bool failFast(TestResult r) // {{{
  {
    return failfast && (r.fail || r.error);
  } // }}}
  void report(TestResult r) // {{{
  {
    final switch (r.status)
    {
      case TestStatus.success: write("."); break;
      case TestStatus.skip:    write("s"); break;
      case TestStatus.fail:    write("!"); break;
      case TestStatus.error:   write("X"); break;
    }
    if (verbose)
      writeln(" ", r.testFile);
    else
      stdout.flush();
  } // }}}
  int diffAndPatch(TestResult r) // {{{
  {
    // if it's closed it should be empty
    if (!r.diff.isOpen)
      return 0;

    if (!nodiffs)
      r.diff.byLine.each!writeln;

    if (!update)
      return 0;

    // $DRAM_UPDATE is either "1", "[Yy]es" or "[Aa]sk"

    if (ask) {
      writefln("%s failed. Apply results? [Y/n]", r.testFile);

      auto reply = std.stdio.stdin.readln;
      if (reply.length && 0 > "Yy".indexOf(reply[0]))
        return 0;
    }

    return runPatch(this, r.testFile, r.diff.name);
  } // }}}
  string[string] env(string test, string tmpdir) // {{{
  {
    string[string] env;
    if (keepenv) {
      // -E / DRAM_KEEP_ENVIRON=[1Yy] preserves existing env. variables
      // DRAM_* ones are always removed, because not doing so would
      // break running dram in dram tests.
      env = environment
        .toAA
        .byPair
        .filter!(nv => !nv.key.startsWith("DRAM_"))
        .assocArray
      ;
    } else {
      // this is the default environment
      env = [
        "COLUMNS": "80"
      , "LANG": "C"
      , "LC_ALL": "C"
      , "LINES": "20"
      , "PATH": environment.get("PATH", "/usr/local/bin:/usr/local/sbin:/bin:/sbin:/usr/bin:/usr/sbin")
      , "TZ": "UTC"
      ];
      // variables forced using -e / $DRAM_ENV
      envvars.each!((n) {
        auto v = environment.get(n);
        if (v != null)
          env[n] = v;
      });
    }
    // these are always present, and specific to each testfile
    [
      "TESTDIR": test.dirName.absolutePath.buildNormalizedPath
    , "TESTFILE": test.baseName
    , "TMPDIR": tmpdir.absolutePath
    ].each!((n, v) => env[n] = v);
    return env;
  } // }}}

  bool matchLine(string expected, string actual) // {{{
  {
    auto el = expected.length;
    if (expected.endsWith(suffixes.glob))
      return globMatch(actual, expected.dropBack(suffixes.glob.length));
    if (expected.endsWith(suffixes.regex))
      return !matchFirst(actual, "^" ~ expected.dropBack(suffixes.regex.length) ~ "$").empty;
    return expected == actual;
  } // }}}
  string inputLine(ulong i, string ln) // {{{
  {
    return (i == 0 ? prompts.ps1 : prompts.ps2) ~ ln;
  } // }}}
  string outputLine(string ln) // {{{
  {
    return indent ~ ln;
  } // }}}
  string exitLine(int exitcode) // {{{
  {
    return format("%s[%d]", indent, exitcode);
  } // }}}
  string cookieLine(int i) // {{{
  {
    return format("echo %s %d $?", cookie, i);
  } // }}}
} // }}}

TestResult runTest(DramConfig cfg, string testFile) // {{{
{
  auto stopWatch = StopWatch(AutoStart.yes);
  auto p = new DramParser(cfg);
  auto asserts = p.parseFile(testFile);

  // each test runs in a private directory, its basename
  // is derived from the absolute pathname of the testfile
  // this is to avoid an annoyance from cram where if you
  // have foo/some.t and bar/some.t, you'll get (roughly)
  // /tmp/cram-$PID/some.t-1 and /tmp/cram-$PID/some.t-2.
  // IOW, it's hard to tell which one is which.
  auto twd = cfg.tmpdir.buildNormalizedPath(
    testFile.absolutePath.buildNormalizedPath.stripLeft("/")
  ).to!string;
  twd.mkdirRecurse;

  // each test work area contains the following files:
  // "script" - the code extracted from the testfile
  // "out" - this will contain the stdout and stderr
  //   resulting from a run of "script"
  // "result" - outputs from "out" merged with the original
  //   testfile
  // "diff" - unified diff between the testfile and "result"
  //
  // "script" runs in `twd`/work.
  auto outputFile = File(twd.buildPath("out"), "w+");
  auto script = File(twd.buildPath("script"), "w+");

  writeScript(cfg, script, asserts);

  script.flush;

  // exit code of 80 is special, it causes dram to ignore
  // outputFile from the script and treat the test as "skipped".
  auto exit = runScript(cfg, testFile, script.name, outputFile);
  if (exit) {
    stopWatch.stop;
    return TestResult(
      exit == 80 ? TestStatus.skip : TestStatus.error
    , testFile
    , twd
    , stopWatch.peek
    );
  }

  outputFile.rewind;

  auto result = File(twd.buildPath("result"), "w+");
  writeResults(cfg, result, asserts, collectOutputs(cfg, outputFile));
  outputFile.close;

  result.close;
  auto diff = File(twd.buildPath("diff"), "w+");
  auto exDiff = runDiff(cfg, testFile, result.name, diff);

  // zero exit from diff means the files are the same,
  // in that case diffAndPatch doesn't need the file.
  if (exDiff)
    diff.rewind;
  else
    diff.close;
  stopWatch.stop;
  return TestResult(
    exit == 0 && exDiff == 0
    ? TestStatus.success
    : exit == 0 && exDiff == 1
      ? TestStatus.fail
      : TestStatus.error
  , testFile
  , twd
  , stopWatch.peek
  , diff
  );
} // }}}

struct TestResult // {{{
{
  TestStatus status;
  string testFile;
  string twd;
  Duration time;
  File diff;
  bool error() // {{{
  {
    return status == TestStatus.error;
  } // }}}
  bool fail() // {{{
  {
    return status == TestStatus.fail;
  } // }}}
  bool skip() // {{{
  {
    return status == TestStatus.skip;
  } // }}}
  bool success() // {{{
  {
    return status == TestStatus.success;
  } // }}}
} // }}}

enum TestStatus // {{{
{
  success
, skip
, fail
, error
} // }}}

// a testfile is prose interspersed with command blocks
// (or vice versa, depending on one's point of view).
// another way to look at it is that it's a series of
// command blocks, where each block consists of a command,
// its outputs, and a prose trailer.
// to account for any leading prose at the beginning of
// the testfile, dram synthesizes an always-successful
// command and skips it when writing out the results.

struct Assertion // {{{
{
  Array!string program;
  Array!string output;
  uint exitcode;
  Array!string trailer;
  this(string cmd)
  {
    program.insertBack(cmd);
  }
} // }}}

class DramParser // {{{
{
  DramConfig cfg;

  Array!Assertion asserts;

  alias LineParser = bool delegate(int, string);
  LineParser[] Start;
  LineParser[] CmdLine;
  LineParser[] OutputLine;
  LineParser[]* parsers;

  Regex!char reExitCode;

  this(DramConfig cfg) // {{{
  {
    this.cfg = cfg;
    reExitCode = regex("^" ~ escaper(cfg.indent).to!string ~ r"\[(\d+)\]");
    parsers    = &Start;
    Start      = [&cmdline, &trailer];
    CmdLine    = [&cmdline, &inputline, &exitcode, &outputline, &trailer];
    OutputLine = [&cmdline, &exitcode, &outputline, &trailer];
    asserts.insertBack(Assertion(":"));
  } // }}}
  ref Assertion current() // {{{
  {
    return asserts.back();
  } // }}}
  Array!Assertion parseFile(string testFile) // {{{
  {
    File(testFile)
      .byLineCopy
      .enumerate(1)
      .each!((nr, ln) => parseLine(nr, ln))
    ;
    return asserts;
  } // }}}
  void parseLine(int lnno, string ln) // {{{
  {
    foreach (parse; *parsers)
      if (parse(lnno, ln)) return;
    assert(0, "parser bug, trailer() is supposed to consume anything");
  } // }}}
  bool cmdline(int lnno, string ln) // {{{
  {
    if (!ln.startsWith(cfg.prompts.ps1))
      return false;
    asserts.insertBack(Assertion());
    current.program ~= ln[cfg.prompts.ps1.length..$];
    parsers = &CmdLine;
    return true;
  } // }}}
  bool inputline(int lnno, string ln) // {{{
  {
    if (!ln.startsWith(cfg.prompts.ps2))
      return false;
    current.program ~= ln[cfg.prompts.ps2.length..$];
    parsers = &CmdLine;
    return true;
  } // }}}
  bool outputline(int lnno, string ln) // {{{
  {
    if (!ln.startsWith(cfg.indent))
      return false;
    ln = ln[cfg.indent.length..$];
    glob(ln) || re(ln) || literal(ln);
    parsers = &OutputLine;
    return true;
  } // }}}
  bool glob(string ln) // {{{
  {
    if (!ln.endsWith(cfg.suffixes.glob))
      return false;
    current.output ~= ln;
    return true;
  } // }}}
  bool re(string ln) // {{{
  {
    if (!ln.endsWith(cfg.suffixes.regex))
      return false;
    current.output ~= ln;
    return true;
  } // }}}
  bool literal(string ln) // {{{
  {
    current.output ~= ln;
    return true;
  } // }}}
  bool exitcode(int lnno, string ln) // {{{
  {
    auto m = matchFirst(ln, reExitCode);
    if (m.empty)
      return false;
    current.exitcode = m[1].to!uint;
    parsers = &Start;
    return true;
  } // }}}
  bool trailer(int lnno, string ln) // {{{
  {
    current.trailer ~= ln;
    parsers = &Start;
    return true;
  } // }}}
} // }}}

void writeScript(DramConfig cfg, File script, Array!Assertion asserts) // {{{
// the outputs from `echo DRAM ...` are used to gather
// exit codes from test code.
{
  asserts[].enumerate(1).each!((int nr, Assertion ass) {
    ass.program.each!(t => script.writeln(t));
    script.writefln(cfg.cookieLine(nr));
  });
} // }}}

int runScript(DramConfig cfg, string test, string script, File outputFile) // {{{
// stdin is redirected from /dev/null
// stdout, stderr from the script are redirected into `outputFile`
// working directory is a freshly created "work", a sibling of
// the script
{
  auto workdir = script.dirName.buildPath("work");
  auto tmpdir  = script.dirName.buildPath("tmp");
  workdir.mkdir;
  tmpdir.mkdir;

  return spawnProcess(
    [cfg.shell, script]
  , File("/dev/null", "r")
  , outputFile
  , outputFile
  , cfg.env(test, tmpdir)
  , Config.retainStderr | Config.retainStdout | Config.newEnv
  , workdir
  ).wait;
} // }}}

Array!Output collectOutputs(DramConfig cfg, File outputFile) // {{{
{
  auto outputs = Array!Output();
  auto o = Output();
  outputFile
    .byLineCopy
    .each!((ln) {
      // the cookie is supposed to be at the beginning of a line
      // if a command output ends without a trailing newline,
      // the cookie line will butt the output.
      auto parts = ln.findSplitBefore(cfg.cookie ~ " ");
      auto noeol = !parts[0].empty && !parts[1].empty;
      if (!parts[0].empty || parts[1].empty)
        o.output ~= parts[0] ~ (noeol ? cfg.suffixes.noeol : "");
      if (!parts[1].empty) {
        o.exitcode = parts[1].drop(1 + parts[1].lastIndexOf(' ')).to!uint;
        outputs ~= o;
        o = Output();
      }
    });
  return outputs;
} // }}}

void writeResults(DramConfig cfg, File sink, Array!Assertion asserts, Array!Output outputs) // {{{
// produce a copy of the testfile with actual outputs merged in
{
  // leading prose (this is the command synthesized so we
  // have something to hang the opening "trailer" on.
  asserts[0].trailer
    .each!(x => sink.writeln(x))
  ;
  zip(asserts[1..$], outputs[1..$]).each!((ass, o) {
    // command
    ass.program[]
      .each!((i, ln) => sink.writeln(cfg.inputLine(i, ln)))
    ;
    // outputs
    diff!((l, r) => cfg.matchLine(l, r))(ass.output, o.output)
      .filter!(edit => edit.newline)
      .each!(edit => sink.writeln(cfg.outputLine(edit.text)))
    ;
    // exit code
    if (o.exitcode)
      sink.writeln(cfg.exitLine(o.exitcode));
    ass.trailer
      .each!(x => sink.writeln(x))
    ;
  });
} // }}}

struct Output // {{{
{
  Array!string output;
  uint exitcode;
} // }}}

int runDiff(DramConfig cfg, string test, string result, File sink) // {{{
{
  auto argv = [cfg.diffcmd, "-u", "--label", test, "--label", test, test, result];
  auto ex = spawnProcess(
    argv
  , std.stdio.stdin
  , sink
  , std.stdio.stderr
  , null
  , Config.retainStdout
  ).wait;
  if (ex > 1)
    throw new ToolFailure(argv);
  return ex;
} // }}}

int runPatch(DramConfig cfg, string test, string patch) // {{{
{
  auto argv = [cfg.patchcmd, "-s", test, patch];
  auto ex = spawnProcess(argv).wait;
  if (ex)
    throw new ToolFailure(argv);
  return ex;
} // }}}

class ToolFailure : Exception
{
  this(string [] argv, string file = __FILE__, size_t line = __LINE__)
  {
    super(argv.map!(quote).join(" "), file, line);
  }
  static auto quote(string s)
  {
    return "'" ~ s.replace("'", "'\\''") ~ "'";
  }
}
